src/Symfony/Component/Security/Http/Firewall/SessionExpirationListener.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\Firewall;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\Security\Core\Exception\SessionExpiredException;
+use Symfony\Component\Security\Http\HttpUtils;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+
+/**
+ * SessionExpirationListener controls idle sessions.
+ *
+ * @author Antonio J. García Lagar <aj@garcialagar.es>
+ */
+class SessionExpirationListener implements ListenerInterface
+{
+    private $tokenStorage;
+    private $httpUtils;
+    private $maxIdleTime;
+    private $targetUrl;
+    private $logger;
+
+    public function __construct(TokenStorageInterface $tokenStorage, HttpUtils $httpUtils, $maxIdleTime, $targetUrl = null, LoggerInterface $logger = null)
+    {
+        $this->tokenStorage = $tokenStorage;
+        $this->httpUtils = $httpUtils;
+        $this->setMaxIdleTime($maxIdleTime);
+        $this->targetUrl = $targetUrl;
+        $this->logger = $logger;
+    }
+
+    /**
+     * Handles expired sessions.
+     *
+     * @param GetResponseEvent $event A GetResponseEvent instance
+     *
+     * @throws SessionExpiredException If the session has expired
+     */
+    public function handle(GetResponseEvent $event)
+    {
+        $request = $event->getRequest();
+        $session = $request->getSession();
+
+        if (null === $session || null === $token = $this->tokenStorage->getToken()) {
+            return;
+        }
+
+        if (!$this->hasSessionExpired($session)) {
+            return;
+        }
+
+        if (null !== $this->logger) {
+            $this->logger->info(sprintf("Expired session detected for user named '%s'", $token->getUsername()));
+        }
+
+        $this->tokenStorage->setToken(null);
+        $session->invalidate();
```

Comment:
Ok, so this is the trickiest part of this PR. You're effectively imitating what `LogoutListener` does + some of its logout handlers (e.g. `SessionLogoutHandler`). So then (to play devil's advocate), why not actually inject the logout handlers and use them here? We *are* logging the user out... so shouldn't those handlers be called? And to go further, couldn't we use the normal `DefaultLogoutSuccessHandler` to figure out where to take the user next?

The disadvantage to this is that you couldn't do something different based on a normal user logout versus a session expiration logout. But, is that a problem? And if it is, could we dispatch a special event (e.g. `security.session_expired`) in this class, *before* calling the logout handlers?

Reply:
After 3 years, my vision is slightly different. I'll remove the `$tokenStorage->setToken(null)` so this listener will only detect the session expiration and optionally (enabled by default) will invalidate the session.

For me, the logout is an interactive user action and different from the session expiration case.
If you need to force the logout you could disable the session invalidation and redirect the user to the logout path. What do you think?

Reply:
Hmm... so then the purpose of this listener would be to *only* invalidate the session after a certain amount of idle time. If that's the case... then this is kind of unrelated to security, right? This could just be a general framework feature of the session: a framework-level way to invalidate sessions that are idle. Wdyt?

Reply:
@weaverryan It took me some time to review this, as the history behind this PR is too old. It is the result of an split of #12009, which was a reimplementation of #786, into #12807(this PR) and #12810, so I decided to write a recap

We have to implement 2 different things here: first, a feature to detect idle sessions, and last, to provide a default behavior once the expired session is detected.

**Detecting idle sessions**

It is now implemented as a firewall listener, so it can be enabled or disabled for different firewalls in the same app, but if the session is invalidated inside one firewall, it will logout from any firewall that uses the `ContextListener` to restore the token.

If we decide that this feature is not related to security, but to the framework, we can implement it as a `kernel.request` event listener that, IMO, should be executed with higher priority than the firewall one, so the user is not authenticated from an expired session. I'm not sure which component (HttpKernel?) should provide this listener.

Anyway, once the expired session is detected, I think that a new event should be dispatched (currently not implemented). Your proposed name `security.session_expired` it's only an option if we decide that this feature is a security concern.

**Default behavior on expired sessions**

For me, the best option here is to invalidate the session (should we allow not to override this?), and redirect the user to a given path that will return an accurate response depending on the given request (html, json, etc...) or to throw a `SessionExpiredException` in no path is defined. If we implement this feature as a kernel event listener, this behavior should be hooked to the new `session_expired` event.

In a previous comment you proposed to force an user logout calling the `LogoutSuccessHandler` and all defined `LogoutHandlers`, but I see two problems here:
  1. In the app there could be more than one firewall, so; which logout should be executed when an expired session is detected if we decide that this option should not be implemented as a firewall listener, but as a kernel listener?
  2. If we decide to execute the logout from the current firewall, (in this case the expired session listener should be a firewall listener, or has to be executed after the firewall listener), what should be done if there is no logout listener defined for the current path?

Reply:
Hey @ajgarlag!

Sorry for the slow reply again :p.

After reading your reply, it seems to me that this PR indeed has nothing to do with security. We're simply implementing a framework-level "session timeout" feature. This is necessary because there's no reliable way to do this with php.ini configuration. If there were, then we would just have the user do that.

With that in mind, this 100% seems like "session" configuration. And so, it would be configured under the `framework.session` configuration. And I don't think ANY other configuration is necessary - no need to have `expiration_url`. Nope, we would simply invalidate the user's session and that's it (e.g. allow the request to continue).

IF we want to allow the user to hook into this further, then dispatching the event we discussed may make sense. I WOULD do that, then we can see what others think.

To summarize (and tell me if you disagree):

1) This is not related to security at all - implement it under the `framework.session` configuration. The listener would be configured in FrameworkBundle, but the listener/subscriber class itself would live in the HttpKernel component.

2) When a session "expires", simply invalidate the session and do nothing else. Well, let's also dispatch a new event, and we can see if people think it's a good idea.